Implemented recipients per basket

lars-devs opened the pull request:
"RecipientsPerItem" may be used in config to define by-basket recipients. Invalid JSON, item to be notified not found in JSON or not defining RecipientsPerItem falls back to regular Recipient as email address that gets notified.

<!-- markdownlint-disable-next-line MD041 -->
### Pull Request Checklist

* [x] Have you checked to ensure there aren't other open
[Pull Requests](../../pulls) for the same update/change?
* [x] Did you make your Pull Request on the dev branch?
* [x] Does your submission pass tests? `make test`
* [x] Have you lint your code locally prior to submission? `make lint`
* [x] Could you build and run the docker image successfully? `make images`
* [x] Could you create a running executable? `make executable`
* [x] Have you added an explanation of what your changes do
and why you'd like to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran manual tests with your changes locally?

Der-Henning commented on config.sample.ini:
```diff
@@ -106,6 +107,7 @@ TLS = true
 SSL = false
 Sender = max.mustermann@gmail.com
 Recipients = max.mustermann@gmail.com
+; RecipientsPerItem = {"123" : ["max.mustermann@example.com", "moritz.mustermann@example.com"], "456" : ["max.mustermann@example.com"]}
```
What about allowing a single recipient to be a string?
In the code you could check if the value is a list and if not pack it into a single item list.

````python
if not isinstance(value, list):
  value = [value]
````

lars-devs replied:
That's a good suggestion, especially for those, who only plan to notify single mail addresses for each item.

Der-Henning commented on tgtg_scanner/notifiers/smtp.py:
```diff
@@ -79,29 +81,61 @@ def _stay_connected(self) -> None:
+
+        # Determine recipient(s) based on item_id
+        if self.recipients_per_item:
+            try:
```
At this point, the configuration should be valid and no longer need to be checked. It is also unnecessary to check the configuration every time before sending, as it cannot change in the meantime. Therefore, please implement an appropriate check somewhere below line 37. Simultaneously, the configuration should be transferred into a variable with appropriate typing.

suggestion (untested):
````python
self.item_recipients: Union[dict[str, list[str]], None] = None
if config.smtp.recipients_per_item is not None:
  json_data = json.loads(config.smtp.recipients_per_item)
    if not isinstance(json_data, dict) or any(not isinstance(value, list) for value in json_data.values):
      raise SMTPConfigurationError("Recipients per Item is not a valid dictionary")
   self.item_recipients = json_data
````

lars-devs replied:
Thank's for Yor feedback! I moved the syntax validation alongside the check for list and string format of mail addresses to the `init()` method below the check for `self.enabled`. The functionality of the method for sending mail only determines the recipient(s) depending on the given item, from now on.

Der-Henning commented on tgtg_scanner/models/config.py:
```diff
@@ -244,6 +244,7 @@ class SMTPConfig(NotifierConfig):
     use_ssl: bool = False
     sender: Union[str, None] = None
     recipients: list[str] = field(default_factory=list)
```
Maybe allow no recipients if recipients per item is set?

lars-devs replied:
I’d stick to `Recipients` being mandatory to have a fallback just in case. Checking the favorited items against the `RecipientsPerBasket` Json at the time of reading configuration wouldn’t be sufficient, since during runtime, a „wild unmatched“ item could appear (because favorites are updated every iteration). Handling this could either mean printing an exception or exiting the application with a non-zero status. Both would not send any mail and isn’t an appropriate solution, in my opinion.

Der-Henning reviewed (commented):
Thank you for your contribution!
This is a good add-on to the current smtp notifier and could be a thing for the other notifiers too. 
Please try to stick to the single responsibility principle, which says that every function or method should only have a single responsibility. 
The responsibility of the _send_mail method is to send a mail. It shouldn't have to check if everything is set up correctly. This is the responsibility of the __init__ method.

lars-devs commented:
Thank You for Your feedback and suggestions! You're right in regards of separation of concerns, that makes it a lot easier to maintain and more secure during runtime. I moved the validation part to the init() method and removed unnecessary parts of the recipient determination. I also updated the config sample to show, that (single) recipients are not limited to `[list representation]` and validated this as well. Still, I prefer the `Recipients` to be mandatory, to make sure, a recipient can always be determined if a favorite gets added during runtime which is not present in the Json.

Covered syntax cases

Fallback to Recipients:
- `RecipientsPerItem` not in configuration
- `RecipientsPerItem = {}`
- `RecipientsPerItem = {"other-id": "a"}` (Item ID not in Json)

Catched by exception:
- `RecipientsPerItem = {{}` (Invalid syntax)
- `RecipientsPerItem = {"id": 000}` (Recipients not list or str format)

Valid formats:
- `RecipientsPerItem = {"id": ["a", "b"]}` (Multiple recipients as list)
- `RecipientsPerItem = {"id": [a]}` (Single recipient as list)
- `RecipientsPerItem = {"id": "a"}` (Single recipient as str)

Der-Henning commented on tgtg_scanner/notifiers/smtp.py:
```diff
@@ -79,29 +96,48 @@ def _stay_connected(self) -> None:
             self._connect()
 
-    def _send_mail(self, subject: str, html: str) -> None:
+    def _send_mail(self, subject: str, html: str, item_id: int) -> None:
```
The item_id should already be of type `str`. But this is an issue with bad typing in models/item.py and a separate task for further refinement.
No change required.

Der-Henning commented on tgtg_scanner/notifiers/smtp.py:
```diff
@@ -79,29 +96,48 @@ def _stay_connected(self) -> None:
+        recipients = self.recipients
+
+        # Determine recipient(s) based on item_id
+        if self.item_recipients:
```
This is unnecessary as `self.item_recipients` should always be a dict at this point.

Der-Henning commented on tgtg_scanner/notifiers/smtp.py:
```diff
@@ -44,6 +46,21 @@ def __init__(self, config: Config, reservations: Reservations, favorites: Favori
+                    raise SMTPConfigurationError("Recipients per Item is not a valid dictionary")
+                self.item_recipients = item_recipients
+            else:
+                self.item_recipients = {}
```
Maybe move this part to the top before line 37 as the default value.

```suggestion
                self.item_recipients: dict[str, list[str]] = {}
```

Der-Henning commented on tgtg_scanner/notifiers/smtp.py:
```diff
@@ -44,6 +46,21 @@ def __init__(self, config: Config, reservations: Reservations, favorites: Favori
+                # and SMTPConfigurationError during key-value validation of item-recipients pairs inside
+                except Exception:
+                    raise SMTPConfigurationError("Recipients per Item is not a valid dictionary")
+                self.item_recipients = item_recipients
```
This ways type `dict[str, list[str]]` is ensured.
```suggestion
                self.item_recipients = {k: v if isinstance(v, list) else [v] for k, v in ite,_recipients.items()}
```

Der-Henning commented on tgtg_scanner/notifiers/smtp.py:
```diff
@@ -79,29 +96,48 @@ def _stay_connected(self) -> None:
+
+        # Determine recipient(s) based on item_id
+        if self.item_recipients:
+            item = str(item_id)
+            if item in self.item_recipients:
+                # Find addresses based on item to be notified
+                recipients = []
+                recipients_for_item = self.item_recipients[item]
+                if isinstance(recipients_for_item, list):
+                    for address in recipients_for_item:
+                        recipients.append(address)
+                else:
+                    address = recipients_for_item
+                    recipients.append(address)
```
if type `dict[str, list[str]]` is ensured you can simply do

```suggestion
            recipients = self.item_recipients.get(str(item_id), self.recipients)
```
This way you could also delete line 110.

The result is a very pythonic one-liner.

Der-Henning reviewed (commented):
In the first review I forgot to mention, that the configuration documentation in wiki/configuration.md should also be modified. The markdown files in wiki/ are uploaded to the github wiki.

Der-Henning commented on tgtg_scanner/notifiers/smtp.py:
```diff
@@ -29,6 +30,8 @@ def __init__(self, config: Config, reservations: Reservations, favorites: Favori
         self.password = config.smtp.password
         self.sender = config.smtp.sender
         self.recipients = config.smtp.recipients
+        self.recipients_per_item = config.smtp.recipients_per_item
```
`self.recipients_per_item` is never used. This line can be deleted.

```suggestion
```

Der-Henning commented:
Let me know when you are done. Then I will merge.

lars-devs commented:
I finally did some testing again and am ready to merge.
